### DISCUSSION TITLE
Punchy 0.0.3 Release Planning

### DISCUSSION THREAD

**Discussion by zicklag** [fishfight/punchy#91] [Planning]
Hey folks! I attempted to do some issue organization and labeling. I'm not sure if all the labels I created will end up being more burdensome than helpful, but we can try it out and if they bug us we'll just remove them.

I tried organize some of the issues into the next Punchy [0.3 milestone](https://github.com/fishfight/punchy/milestone/3), but I'm not sure if all of those tasks are something we want to do before the next release.

Specifically for #27 ( items ) and #57 ( big boss ), I felt like maybe making the Punchy 0.3 release a little smaller in favor of getting what we've gotten so far out, and doing those in Punchy 0.4. Or maybe we could just pick one item to do.

At the same time, I'm not sure what else @odecay and @64kramsystem are going to want to be working on next or how hard they think those will be, so if those aren't going to be a big deal to do, we could just do them.

I feel like frequent small releases gives a nice feeling of progress. :grin:

**Comment by 64kramsystem** [comment]
I like labels, especially when they're prefixed (actually, _only_ when they're prefixed :joy:), so I totally support the new ones :smile: 

Regarding this:

> Specifically for https://github.com/fishfight/punchy/issues/27 ( items ) and https://github.com/fishfight/punchy/issues/57 ( big boss ), I felt like maybe making the Punchy 0.3 release a little smaller in favor of getting what we've gotten so far out, and doing those in Punchy 0.4. Or maybe we could just pick one item to do.

For large tasks (or multi-task issues), I'd personally use a GitHub project rather than an issue. This allows granular splitting and management. For example, completing the boss encompasses several small sub-tasks: camera handling, AI, moves, etc.etc. By converting it into a project, we could split the tasks into manageable ones.

With "manageable" I mean:

- (more) predictable
- easily distributable across milestones
- granular

Also, since each sub task is an issue:

- a PR can be linked at will automatically close it :grimacing:
- individual tasks can be assigned by standard Github assignment, rather than writing manually into the enclosing issue.

To be 100% clear, what I propose is to use projects as simply collections of issues - nothing more!

**Reply by zicklag** [reply]
Yeah, while tracking issues can work and GitHub has been making them nicer with integrations in the UI, it can be nice to have them not actually "issues".

I'd be good using projects for those kinds of things, but apparently projects are an organization-scoped resource in GitHub, so you have to access to the FishFight org to create them, not just collaborator access on the punchy repository. Not sure if that's a problem or not.

It's a little annoying that there isn't such a think as repo-scoped projects, but I suppose it makes sense since projects are supposed to be able to span repositories.

**Comment by 64kramsystem** [comment]
> At the same time, I'm not sure what else @odecay and @64kramsystem are going to want to be working on next or how hard they think those will be, so if those aren't going to be a big deal to do, we could just do them.

Regarding my next tasks, it's hard to say - right now I'm on holiday and can't find the time :cursing_face::cursing_face:, so unfortunately, I can't predict the next task until I complete any given one :sob: 

After audio, I think I'll do the camera work for the boss section(s).

**Reply by zicklag** [reply]
Yeah, that's totally fine. We don't necessarily need to know what next things are exactly, and any planning doesn't necessarily need to dictate what you do next. You can work on whatever you can find the ability/motivation to work on.

**Comment by 64kramsystem** [comment]
@erlend-sh do you have experience/thoughts on GH projects?

**Reply by erlend-sh** [reply]
I have some experience with GitHub Projects and similar tools & workflows, yeah. There’s a very fine line between helpful and obstructive structure. Every new tool added to the existing workflow brings with it risk of friction because of the additional overhead. But done right, we get greater clarity.

Let’s start with one or two such project boards and see if we can make good use of them. I’ve started the first one here:
https://github.com/fishfight/punchy/projects?type=beta

> It's a little annoying that there isn't such a think as repo-scoped projects, but I suppose it makes sense since projects are supposed to be able to span repositories.

Important to note that this tells us GitHub hasn’t really made Projects with our use case primarily in mind. It’s really intended first and foremost for org-wide initiatives. But it might still work well enough for what we want to do as well.

**Reply by zicklag** [reply]
> here’s a very fine line between helpful and obstructive structure.

Yeah, already I'm feeling like projects are a little "heavy" for what are really simple "Epics" as they've been called in many other project management/agile terms. It feels like tracking issues might still be the lowest friction way to go.

Also, I just realized that it's kind of useful to be able to comment on the tracking issues to help coordinate related work, and projects don't have that.

It seems like they are more suited for release planning maybe ( which milestones are enough for us for now ), or more of a Kanban board style of task management, which we are really all kind of managing on our own terms anyway because we all have different schedules and work plans.

**Reply by 64kramsystem** [reply]
In the use I had in mind, we'd just use projects to split issues that currently would be "scope:large" into small ones - I think this usage is fairly lightweight. It's not great though, that there is no way to comment on a project.

**Comment by erlend-sh** [comment]
Most of the new labels are good! 👍 

I question the usefulness of audience:user. It applies to nearly everything, similar to kind:enhancement. When a label is used practically everywhere, that’s a sign it might be redundant.

**Reply by zicklag** [reply]
Yeah, good point about the redundant labels. I'll take those out.

**Comment by erlend-sh** [comment]
> I feel like frequent small releases gives a nice feeling of progress.

No doubt! We’re gonna try our best to maintain a bi-weekly cadence of releases. The tasks within a milestone are merely intended as a general direction to be moving in; we’re not gonna let unfinished tasks block a release.

> I felt like maybe making the Punchy 0.3 release a little smaller in favor of getting what we've gotten so far out, and doing those in Punchy 0.4.

Yep, that’s what we’ll always do to get a release out. Pending tasks will get relegated to the next release.

However, a task being out of scope is also often a sign that we’re thinking too big for its initial v1 iteration. Items for example already exists in a v1 form as soon as *one* of those sub-tasks we’ve listed is completed.

Boss-man v1 could be a copy of our existing bare-bones enemy AI plus the fact that every so often he’ll do his jump attack, which for now can work just like the dive-attack for players.

**Comment by 64kramsystem** [comment]
Hmmm, I wanted to add an [issue](#97) to a project, but the project doesn't show in the dropdown:

![image](https://user-images.githubusercontent.com/1595356/179321889-fe11e969-0c2b-4ca6-8649-34d4b5b862ba.png)

**Reply by zicklag** [reply]
You can copy the issue URL and paste it into the textbox at the bottom of the window. ( Or anywhere else you may add an issue )

**Reply by 64kramsystem** [reply]
Oh thanks :smile: 

(I've realized that the current project is not a good fit for the issue :grimacing:)

**Comment by 64kramsystem** [comment]
I like the newly introduced tracker issue concept, which solves the problem I have in mind. For me, projects are redundant now :grimacing:.

**Comment by erlend-sh** [comment]
Sounds like that’s a 👎 on the use of Projects for now then.

I had the idea of using Milestones instead, by having both version-milestones and feature-milestones. Sadly an issue can’t belong to more than one milestone at a time.

Tracking issues feels quite sufficient though. We should also pin the most relevant ones, I’ll do that now.

**Reply by zicklag** [reply]
Yeah, at least we tested it out and failed fast. :) 

I feel like tracking issues are sufficient, and good idea to pin them. :+1:

**Comment by zicklag** [comment]
let me know if you guys would like a different approach, but I'm removing the tracking issues from the milestone, in favor of putting only the tracked issues in the milestone.

For instance we have the boss fighter tracking issue ( #57 ), and other issues such as https://github.com/fishfight/punchy/issues/108 and https://github.com/fishfight/punchy/issues/97 that are tasks related to finishing off the boss. I think only the tasks should go in the milestone, because the goal isn't to finish off the entire boss within the milestone, just some of the tasks.

I think it makes it a little easier to track.

Again, just let me know if you think otherwise and we can put them back.

**Reply by 64kramsystem** [reply]
> let me know if you guys would like a different approach, but I'm removing the tracking issues from the milestone, in favor of putting only the tracked issues in the milestone.

That's great! It's actually exactly how I had imagined the role/dynamics of "issue groupers" (whether projects, or tracking issues, or anything else) :smile: :+1:

**Comment by erlend-sh** [comment]
Release is nearly upon us; I see a date was added to honor the aforementioned 14-day cadence. Do keep in mind that any number of days over 14 is perfectly fine, but we will be in release-mode for however long we need to be from that point onwards.

We haven’t done a verbose release post on itch.io yet, so I reckon we’ll do that now. I’ll write the body of the post, but I’d like to request all of you devs @edgarssilva @odecay @Byteron @zicklag @64kramsystem to write in a small piece here about one or a few features/problems that you worked on (with enjoyment, one would hope!). For any version, because this is for our announcement post on itch, not specifically the v0.0.2 release.

Feel free to even delve lightly into how you went about solving something, as long as it’s understandable by novice coders/gamedevs.

Whole new replies for your pieces please, that way we can comment further on each individually in the nested replies.

**Comment by zicklag** [comment]
@erlend-sh Here's my piece for the post you're making. I'm assuming I'm writing it in first person, I can modify it if we want to do something different.

---

The major things I've worked on so far were the UI and the input mapper.

The UI uses the awesome [egui](https://github.com/emilk/egui) library with custom 9-patch style widgets. A fun feature to add was the widget adjacency map, which allowed us to say whether one button was above, below, to the left, or to the right of another button. That way, when you press a direction on your gamepad, the UI can select the next button in that direction on the screen.

The input mapper was a combination of UI work for the settings menu, and a big collaboration with @alice-i-cecile and @cleancut for adding gamepad axis support to the [Leafwing Input Manager](https://github.com/Leafwing-Studios/leafwing-input-manager) Bevy plugin. Now we can support inputs from gamepad joysticks, DPads, and the keyboard without having to handle them individually in the character controller. And because we contributed to the Leafwing Input Manager, other people will be able to use the new features in their own Bevy games!

**Reply by alice-i-cecile** [reply]
LWIM crate owner here: this work rocked, and will be incredibly useful for the Bevy ecosystem as a whole :) Thanks!

**Comment by 64kramsystem** [comment]
List of the most important (and/or (arguably) interesting) tasks:

Basic camera work; it's now activated by the rightmost player push, and limited by the leftmost one, in the style of convenientional beat 'em ups; previously, the camera would just follow the movement of the first player.
In the near future, conditional camera locking will be implenented (e.g. bosses, enemy waves etc.).

Made Github workflows caching shared across all the PRs. Surprisingly, it's not very well known that workflows caching is _not_ shared across PRs, with the result that caching is used only per PR, which causes the first build of any PR to be very slow. I'm actually going to publish a small article shortly, since this is a very important topic.

Added audio base support, with some effects and music. This has been implemented by conventionally choosing the [Bevy Kira Audio](https://github.com/NiklasEi/bevy_kira_audio) plugin, which makes audio management simple (although it's currently doesn't support fade out).

Investigated slowness of build, and ultimately removed any optimization for debug builds, since building the project, even incrementally, was causing a considerable strain on slower setups (e.g. laptops).
While the change itself is trivial, it was very unexpected (and therefore, time consuming) because `opt-level=1` is generally very fast, while, on this project, it turned out to be a few times slower than `opt-level=0`. Interestingly, this was noticed by running the Linux `perf` tool on the cargo build of an incremental build.
On general projects, Rust devs who, for any reason, require `opt-level=1`, but are experiencing a slow build, can disable LTO entirely (`lto=false`) - somewhat surprisingly, a faster variant of LTO ("thin" LTO) is enabled on `opt-level=1`, but it may still take a considerable time!
It's not yet clear what exactly causes the comparative slowness of the project build (although there are educated guesses). In the future, this chapter will be resumed :smile:

**Comment by 64kramsystem** [comment]
Regarding these:

> At the same time, I'm not sure what else @odecay and @64kramsystem are going to want to be working on next

Yesteday I came back from traveling. From tomorrow, I'm back to Real Life™, and I'll be able to get back to work™ :heart_eyes:.

Roughly speaking, besides the simple #104, I'd take #99.

>  or how hard they think those will be, so if those aren't going to be a big deal to do, we could just do them.

The boss is tricky, not because of inherent difficulty, but because I think it requires a careful design.

With ECS, IMO, it's easy to explode the number of components/behaviors, so I think it's important to review the design, to see how it fit, and if minor changes are required.

To be clear, it's perfectly possible that we can fully recycle current behaviors/components. This would be great :smile:, but personally, I can't confirm without looking at the code :smile:

**Reply by erlend-sh** [reply]
That’s why we wanna start off with the no-design approach of #108 to get some first impressions to learn from before getting into deeper design considerations.

It’s still likely to get bumped to v0.0.4 though. We’ll have a quick release discussion during today’s meeting.

**Comment by odecay** [comment]
I just went back and checked played the current web build, we've made so much progress since v0.0.2!
Do we want to make any updates to the readme for v0.0.3 release aside from the video update #90?

**Reply by zicklag** [reply]
I don't know. It seems fairly complete right now. I can't think of anything else to put in it right off :thinking: , but I know what you mean!

It's been a big update, but I think an updated video is the best thing we can do since GitHub doesn't let us embed the web demo as an iframe. :smile: 

Unless we just wanted to move the web link all the way to the top of the README to make it slightly more visible. Something like 🤷‍♂️:

> A 2.5D side-scroller beat-’em-up, made in Bevy. Inspired by games like Little Fighter 2, River City Ransom and [many more](https://fextralife.com/a-history-of-the-side-scrolling-beat-em-up-part-1/). Fully playable in the browser: [Punchy Web Game](https://fishfight.github.io/punchy/demo/).

**Reply by odecay** [reply]
Sounds reasonable, @erlend-sh also referred me to these sections of the FishFight proper readme in discord.
https://github.com/fishfight/FishFight#download--play
https://github.com/fishfight/FishFight#building

**Reply by 64kramsystem** [reply]
It's actually possible to embed videos in the README.md, although not Youtube ones.

It's interesting (and odd), because Github does not store the video in the repository, so there's no large commit (just a link in the README).

I've just tried both adding a new README, and editing an existing one, and it works (via Github web interface).

**Reply by odecay** [reply]
We have the video embeded already, that was in reference to the web build.

**Reply by 64kramsystem** [reply]
> We have the video embeded already, that was in reference to the web build.

ooops, sorry :flushed:  :joy:

**Comment by zicklag** [comment]
I was thinking that it would be good to keep the old web builds around for every release, and then we could link to the release-pinned version of the web demo in the release notes.

That way we have a playable history of the game development. At least until web standards change and break our game. :smile: 

We should automate this in the GitHub pipeline, but I'm busy enough with other stuff right now so if nobody else wants to do that I'd be happy enough just manually pushing the 0.2.0 and 0.3.0 web builds to their own folders in the `gh-pages` branch so that we have them there, then we can automate it for the 0.4.0 release.

**Reply by odecay** [reply]
@zicklag lets do it manually for this one, where should they go?

**Reply by zicklag** [reply]
Since we already have the `/demo` path, it seems like either `/releases/v0.2.0/demo` or `demo/v0.2.0` would be reasonable paths. It doesn't matter all that much.

**Reply by zicklag** [reply]
Oh, I just realized, we might have to build the demo specifically for the path it's hosted under. I'm revisiting the web build script in #165 already so I'll look at that real quick.

**Reply by zicklag** [reply]
Uh oh, for some reason the web build seems to somehow not even get to the Rust main function in release builds now. Not sure how long that's been happening, but I think I'll have to get to that tomorrow.

**Comment by zicklag** [comment]
We got the web build fix and as far as I know, there's no more pending tasks that need to happen before making the release!

**Reply by odecay** [reply]
Sounds good, I'll cut it up.

**Comment by zicklag** [comment]
@odecay could we update the release notes for 0.0.2 and 0.0.3 to include links to the pinned version of the web player for that release?

And would it be good to mention those versions in the README, or do you think that's unnecessary.

**Reply by odecay** [reply]
Ill update and make note that previous releases can be found on their release pages. Would you move the v0.0.2 web build to the correct place if you haven't already?

**Reply by zicklag** [reply]
Yeah, I'll do that.

**Reply by zicklag** [reply]
There we go: https://fishfight.github.io/punchy/player/v0.0.2/

**Reply by odecay** [reply]
updated on release post for v0.0.2 and pointed to release pages in readme for previous versions. (I also changed the link on v0.0.3 to link directly to its version instead of latest)

**Comment by zicklag** [comment]
Awesome! The release is done!

Great job everybody! This is going really well. :D

**Reply by odecay** [reply]
I think all that remains to be done is a release blogpost which @erlend-sh was going to do and linking that in the release for v0.0.3 when its ready.

**Comment by odecay** [comment]
My summary of last release cycle for @erlend-sh blog post.

A [feature](https://github.com/fishfight/punchy/pull/134) I worked on over the past release cycle lays the groundwork for future attack and combo system work which which will be tracked [here](https://github.com/fishfight/punchy/issues/149). I reworked attacks to align with a design commonly used in 2d Fighting Games and 2.5d Beatem-ups. This design is centered around attacks which are split into Startup frames, in which the animation of an attack begins to play, Active frames, where an active collider called a Hitbox is spawned and Recovery frames, in which the Hitbox is removed and the attack animation finishes playing. Having these three stages of attack and defined Hitboxes will allow us to easily make granular design decisions about the way attacks interact with each other, and when characters can be interrupted in an attack or attack string. Upcoming follow up work to this will be to move these attack specific properties to the YAML fighter definitions which will allow for making quick changes to characters attacks, and eventually will be used to specialize character and weapon specific attacks.

I also spent time over these past two weeks doing administrative and planning work like reviewing and merging contributions on github, talking with Artists in our discord, and meeting and conversation around our planned integration with the Ultimate platform, which will be documented in #153 as it progresses.

I noticed toward the end of the release cycle that the contributor workflow for making PRs seemed much improved by both the implementation of bors, and the increasing availability and willingness of contributors to review and merge contributions. I look forward to our continued progress and the planned features and improvements for next release!

**Comment by erlend-sh** [comment]
That is a WRAP on v0.0.3!

https://spicylobster.itch.io/punchy/devlog/409761/punchy-makes-its-debut-
https://github.com/fishfight/punchy/releases/tag/v0.0.3
https://twitter.com/spicylobsterfam/status/1552768660401537024
https://www.reddit.com/r/rust/comments/waltwb/media_fish_fight_punchy_v003_beatemup_made_in/

Amazing work everyone! 👏❤️‍🔥
